batch/batch_test.go

```diff
@@ -0,0 +1,86 @@
+package batch_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/uw-labs/substrate"
+
+	"github.com/uw-labs/substrate-tools/batch"
+	"github.com/uw-labs/substrate-tools/message"
+	"github.com/uw-labs/substrate-tools/mock"
+)
+
+func TestBatchMessageSource(t *testing.T) {
+	t.Parallel()
+	tt := []struct {
+		Name         string
+		Messages     []substrate.Message
+		BatchSize    int
+		HandlerError error
+
+		ExpectedBatches [][]substrate.Message
+		ExpectsError    bool
+	}{
+		{
+			Name: "It should process messages in batches",
+			Messages: []substrate.Message{
+				message.FromString("message-1"),
+				message.FromString("message-2"),
+				message.FromString("message-3"),
+				message.FromString("message-4"),
+			},
+			BatchSize: 2,
+			ExpectedBatches: [][]substrate.Message{{
+				message.FromString("message-1"),
+				message.FromString("message-2"),
+			}, {
+				message.FromString("message-3"),
+				message.FromString("message-4"),
+			}},
+		},
+		{
+			Name: "It should return errors from handler",
+			Messages: []substrate.Message{
+				message.FromString("handler-error"),
+			},
+			BatchSize: 1,
+			ExpectedBatches: [][]substrate.Message{{
+				message.FromString("handler-error"),
+			}},
+			ExpectsError: true,
+			HandlerError: fmt.Errorf("handler-error"),
+		},
+	}
+
+	for _, tc := range tt {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			source := batch.NewMessageSource(
+				&mock.AsyncMessageSource{Messages: tc.Messages},
+				batch.WithBatchSize(tc.BatchSize),
```

Comment:
Looking at the usage, to me it would sound better to remove the middle `batch` word and have `batch.WithSize`.